application_sdk/common/credential_utils.py

```diff
@@ -0,0 +1,130 @@
+"""Utilities for credential providers."""
+
+import collections.abc
+import copy
+import json
+from typing import Any, Dict
+
+from application_sdk.observability.logger_adaptor import get_logger
+
+logger = get_logger(__name__)
+
+
+class CredentialError(Exception):
+    """Base exception for credential-related errors."""
+
+    pass
+
+
+async def resolve_credentials(credentials: Dict[str, Any]) -> Dict[str, Any]:
```

Comment:
everything except this method should be moved to either of
- `error_codes.py`
- `inputs/secretstore.py`